File: backend/user-service/app/repository/user_repository.py
from typing import Optional, List
from uuid import UUID
from sqlalchemy.orm import Session
from app.models.user import User


class UserRepository:
    def __init__(self, db: Session):
        self.db = db

    def find_by_keycloak_id(self, keycloak_id: str) -> Optional[User]:
        return self.db.query(User).filter(User.keycloak_id == keycloak_id).first()

    def find_by_id(self, user_id: UUID) -> Optional[User]:
        return self.db.query(User).filter(User.id == user_id).first()

    def find_by_email(self, email: str) -> Optional[User]:
        return self.db.query(User).filter(User.email == email).first()

    def find_all(self) -> List[User]:
        return self.db.query(User).all()

    def find_athletes(self, fitness_level: str = None, primary_goal: str = None, city: str = None) -> List[User]:
        query = self.db.query(User).filter(User.role == "Athlete", User.profile_complete == True)
        if fitness_level:
            query = query.filter(User.fitness_level == fitness_level)
        if primary_goal:
            query = query.filter(User.primary_goal == primary_goal)
        if city:
            query = query.filter(User.city.ilike(f"%{city}%"))
        return query.all()

    def find_trainers(self) -> List[User]:
        return self.db.query(User).filter(User.role == "Trainer").all()

    def create(self, user: User) -> User:
        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)
        return user

    def update(self, user: User) -> User:
        self.db.commit()
        self.db.refresh(user)
        return user

    def delete(self, user: User) -> None:
        self.db.delete(user)
        self.db.commit()

File: backend/user-service/src/user/user.repository.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from './user.entity';

@Injectable()
export class UserRepository {
  constructor(@InjectRepository(User) private repo: Repository<User>) {}

  findById(id: string) {
    return this.repo.findOne({ where: { id } });
  }

  findByKeycloakId(keycloakId: string) {
    return this.repo.findOne({ where: { keycloak_id: keycloakId } });
  }

  findAll() {
    return this.repo.find();
  }

  findAthletes(fitnessLevel?: string, primaryGoal?: string, city?: string, excludeId?: string) {
    const qb = this.repo.createQueryBuilder('u')
      .where('u.role = :role', { role: 'Athlete' });
    if (excludeId) qb.andWhere('u.id != :excludeId', { excludeId });
    if (fitnessLevel) qb.andWhere('u.fitness_level = :fitnessLevel', { fitnessLevel });
    if (primaryGoal) qb.andWhere('u.primary_goal = :primaryGoal', { primaryGoal });
    if (city) qb.andWhere('u.city ILIKE :city', { city: `%${city}%` });
    return qb.getMany();
  }

  findTrainers() {
    return this.repo.find({ where: { role: 'Trainer' } });
  }

  save(user: Partial<User>) {
    return this.repo.save(user);
  }

  async delete(id: string) {
    await this.repo.delete(id);
  }
}
